### PR TITLE
[Feature] Detect anomalous parameters

### PR DESCRIPTION
**Pull request by jshilong**
## Motivation

Sometimes there are some anomalous parameters in the training phase, mainly two cases 

- Some parameters are  not be used in a forward pass due to our wrong implementation, which causes the error when doing `loss.backward()`
- Some parameters are not be used to produce loss, which causes the deadlock when doing `loss.backward()`
 
In both cases，anomalous parameters are not included in the computational graph that is with `loss` as the root.

## Modification

I add a debug option named `detect_anomalous_params`  to `OptimizerHook`, which can help you find anomalous parameters  

## BC-breaking (Optional)
None

### CODE REVIEW

**Comment by jshilong** [comment]
Ping @ZwwWayne

File: tests/test_runner/test_hooks.py [hunk 1]
```diff
@@ -39,6 +39,85 @@
 sys.modules['petrel_client.client'] = MagicMock()
 
 
+def test_optimizerhook():
+
+    class Model(nn.Module):
```
Comment (ZwwWayne) [inline tests/test_runner/test_hooks.py:44]: The unit test only test the case when conv3 is not used in the forwarwd pass. We should also test the case when conv3 is forwarded but not used for computing loss.

Comment (jshilong) [reply tests/test_runner/test_hooks.py:44]: This case has been tested when the loss is `x1.sum()` then conv2 would be detected. Line 98

File: mmcv/runner/hooks/optimizer.py [hunk 2]
```diff
@@ -20,9 +21,27 @@
 
 @HOOKS.register_module()
 class OptimizerHook(Hook):
+    """A hook contains custom operations for the optimizer.
+
+    Args:
+        grad_clip (dict): A config dict to control the clip_grad.
+            Default: None.
```
Comment (MeowZheng) [inline mmcv/runner/hooks/optimizer.py:28]: (dict) -> (dict, optional)
when the default value is `None`,  the type of keyword argument can be `optional`.

**Review by zhouzaida — APPROVED** [review]

**Comment by zhouzaida** [comment]
After resolving the above comments and the CI passes, the PR can be merged.

**Review by MeowZheng — APPROVED** [review]

**Comment by ZwwWayne** [comment]
Can we check whether it slows down the training speed if we set detect_anomalous_params=True?

**Comment by zhouzaida** [comment]
> Can we check whether it slows down the training speed if we set detect_anomalous_params=True? This PR can be merged after checking the influence on training speed and updating this information in docstring.

It probably slows down the training speed but the flag should only be opened for debugging.

**Comment by zhouzaida** [comment]
Please merge the upstream master branch to this PR which will resolve the error when building documentation.

**Comment by jshilong** [comment]
> > Can we check whether it slows down the training speed if we set detect_anomalous_params=True? This PR can be merged after checking the influence on training speed and updating this information in docstring.
> 
> It probably slows down the training speed but the flag should only be opened for debugging.



> Can we check whether it slows down the training speed if we set detect_anomalous_params=True? This PR can be merged after checking the influence on training speed and updating this information in docstring.

It is only used for debugging

**Comment by jshilong** [comment]
> Can we check whether it slows down the training speed if we set detect_anomalous_params=True? This PR can be merged after checking the influence on training speed and updating this information in docstring.

I will add a notice in the docstr to indicate this is only used for debugging

**Comment by zhouzaida** [comment]
> I will add a notice in the docstr to indicate this is only used for debugging

Hi @jshilong , we will merge the PR after you add a notice in the docstring.

**Review by ZwwWayne — APPROVED** [review]

**Comment by jshilong** [comment]
> > I will add a notice in the docstr to indicate this is only used for debugging
> 
> Hi @jshilong , we will merge the PR after you add a notice in the docstring.

Done

**Review by zhouzaida — APPROVED** [review]
